pkg/controller/bundle/bundle_unpacker.go

```diff
@@ -0,0 +1,388 @@
+		rbacName: "olm-bundle-unpacker",
+		loader:   configmap.NewBundleLoader(),
+	}
+
+	unpacker.apply(options...)
+	if err := unpacker.validate(); err != nil {
+		return nil, err
+	}
+
+	return unpacker, nil
+}
+
+func WithCopyImage(copyImage string) ConfigmapUnpackerOption {
+	return func(unpacker *configmapUnpacker) {
+		unpacker.copyImage = copyImage
+	}
+}
+
+func WithClient(client kubernetes.Interface) ConfigmapUnpackerOption {
+	return func(unpacker *configmapUnpacker) {
+		unpacker.client = client
+	}
+}
+
+func WithConfigMapLister(cmLister listerscorev1.ConfigMapLister) ConfigmapUnpackerOption {
+	return func(unpacker *configmapUnpacker) {
+		unpacker.cmLister = cmLister
+	}
+}
+
+func WithJobLister(jobLister listersbatchv1.JobLister) ConfigmapUnpackerOption {
+	return func(unpacker *configmapUnpacker) {
+		unpacker.jobLister = jobLister
+	}
+}
+
+func WithRoleLister(roleLister listersrbacv1.RoleLister) ConfigmapUnpackerOption {
+	return func(unpacker *configmapUnpacker) {
+		unpacker.roleLister = roleLister
+	}
+}
+
+func WithRoleBindingLister(rbLister listersrbacv1.RoleBindingLister) ConfigmapUnpackerOption {
+	return func(unpacker *configmapUnpacker) {
+		unpacker.rbLister = rbLister
+	}
+}
+
+func WithNow(now func() metav1.Time) ConfigmapUnpackerOption {
+	return func(unpacker *configmapUnpacker) {
+		unpacker.now = now
+	}
+}
+
+func (c *configmapUnpacker) apply(options ...ConfigmapUnpackerOption) {
+	for _, option := range options {
+		option(c)
+	}
+}
+
+func (c *configmapUnpacker) validate() (err error) {
+	// TODO: Better validation
+	switch {
+	case c.copyImage == "":
+		err = fmt.Errorf("no copy image given")
+	case c.client == nil:
+		err = fmt.Errorf("client is nil")
+	case c.cmLister == nil:
+		err = fmt.Errorf("configmap lister is nil")
+	case c.jobLister == nil:
+		err = fmt.Errorf("job lister is nil")
+	case c.roleLister == nil:
+		err = fmt.Errorf("role lister is nil")
+	case c.rbLister == nil:
+		err = fmt.Errorf("rolebinding lister is nil")
+	case c.loader == nil:
+		err = fmt.Errorf("bundle loader is nil")
+	case c.now == nil:
+		err = fmt.Errorf("now func is nil")
+	}
+
+	return
+}
+
+func (c *configmapUnpacker) UnpackBundle(lookup *operatorsv1alpha1.BundleLookup) (result *BundleUnpackResult, err error) {
+	result = newBundleUnpackResult(lookup)
+	cm, err := c.ensureConfigmap(result.CatalogSourceRef.Namespace, result.name)
+	if err != nil {
+		return
+	}
+
+	var role *rbacv1.Role
+	role, err = c.ensureRole(cm.GetNamespace(), cm.GetName())
+	if err != nil {
+		return
+	}
+
+	_, err = c.ensureRoleBinding(role.GetNamespace())
+	if err != nil {
+		return
+	}
+
+	var job *batchv1.Job
+	job, err = c.ensureJob(cm.GetNamespace(), cm.GetName(), result.Path)
+	if err != nil {
+		return
+	}
+
+	cond := result.GetCondition(operatorsv1alpha1.BundleLookupPending)
+	now := c.now()
+	if job == nil || len(job.Status.Conditions) == 0 || len(job.Status.Conditions) > 0 && job.Status.Conditions[0].Type != batchv1.JobComplete {
```

Comment:
non-blocking: there may be some nice ways to clean this up a bit, e.g. `job.Status.ConditionMatches(operatorsv1alpha1.BundleLookupPending, func(Condition) bool { } )`